docs/source/best_practices.rst

```diff
@@ -0,0 +1,187 @@
+and maintain each piece of the code. Note that LibCST is traversing the tree completely
+in both cases, so while the first appears to be faster, it is actually doing the same
+amount of work as the second.
+
+Prefer ``updated_node`` when modifying trees
+--------------------------------------------
+
+When you are using :ref:`libcst-visitors` to modify a LibCST tree, only return
+modifications to ``updated_node``. The ``original_node`` parameter on any
+``leave_<Node>`` method is provided for book-keeping and is guaranteed to be
+equal via ``==`` and ``is`` checks to the ``node`` parameter in the corresponding
+``visit_<Node>`` method. Remember that LibCST trees are immutable, so the only
+way to make a modification is to return a new tree. Hence, by the time we get to
+calling ``leave_<Node>`` methods, we have an updated tree whose children have been
+modified. Therefore, you should only return ``original_node`` when you want to
+explicitly discard changes performed on the node's children.
+
+Say you wanted to rename all function calls which were calling global functions.
+So, you write the following::
+
+    class FunctionRenamer(cst.CSTTransformer):
+        def leave_Call(self, original_node: cst.Call, updated_node: cst.Call) -> cst.Call:
+            if m.matches(original_node.func, m.Name()):
+                return original_node.with_changes(
+                    func=cst.Name(
+                        "renamed_" + cst.ensure_type(original_node.func, cst.Name).value
+                    )
+                )
+            return original_node
+
+Consider writing instead::
+
+    class FunctionRenamer(cst.CSTTransformer):
+        def leave_Call(self, original_node: cst.Call, updated_node: cst.Call) -> cst.Call:
+            if m.matches(updated_node.func, m.Name()):
+                return updated_node.with_changes(
+                    func=cst.Name(
+                        "renamed_" + cst.ensure_type(updated_node.func, cst.Name).value
+                    )
+                )
+            return updated_node
+
+The version that returns modifications to ``original_node`` has a subtle bug. Consider
+the following code snippet::
+
+    some_func(1, 2, other_func(3))
+
+Running the recommended transform will return us a new code snippet that looks like this::
+
+    renamed_some_func(1, 2, renamed_other_func(3))
+
+However, running the version which modifies ``original_node`` will instead return::
+
+    renamed_some_func(1, 2, other_func(3))
+
+That's because the ``updated_node`` tree contains the modification to ``other_func``.
+By returning modifications to ``original_node`` instead of ``updated_node``, we accidentally
+discarded all the work done deeper in the tree.
+
+Provide a ``config`` when generating code from templates
+--------------------------------------------------------
+
+When generating complex trees it is often far easier to pass a f-string to
```

Comment:
Why it needs to be a f-string?

Reply:
Because it primes the idea of using the various parsers to parse templated code.